File: internal/store/schema.go
package store

import (
	"database/sql"
	"fmt"

	_ "modernc.org/sqlite"
)

const schemaSQL = `
CREATE TABLE IF NOT EXISTS domain_keys (
    domain      TEXT NOT NULL,
    key_type    TEXT NOT NULL,
    algorithm   TEXT NOT NULL,
    public_key  BLOB NOT NULL,
    private_key BLOB,
    key_salt    BLOB,
    key_nonce   BLOB,
    key_id      TEXT NOT NULL,
    created_at  TEXT NOT NULL,
    expires_at  TEXT NOT NULL,
    revoked_at  TEXT,
    revocation_reason  TEXT,
    replacement_key_id TEXT,
    PRIMARY KEY (domain, key_type)
);

CREATE TABLE IF NOT EXISTS user_keys (
    address     TEXT NOT NULL,
    key_type    TEXT NOT NULL,
    algorithm   TEXT NOT NULL,
    public_key  BLOB NOT NULL,
    private_key BLOB,
    key_salt    BLOB,
    key_nonce   BLOB,
    key_id      TEXT NOT NULL,
    created_at  TEXT NOT NULL,
    expires_at  TEXT NOT NULL,
    revoked_at  TEXT,
    revocation_reason  TEXT,
    replacement_key_id TEXT,
    PRIMARY KEY (address, key_type, key_id)
);

CREATE TABLE IF NOT EXISTS device_certificates (
    device_key_id       TEXT NOT NULL PRIMARY KEY,
    user_id             TEXT NOT NULL,
    device_id           TEXT NOT NULL,
    issuing_device_key_id TEXT NOT NULL,
    scope_json          TEXT NOT NULL,
    issued_at           TEXT NOT NULL,
    expires_at          TEXT,
    signature_json      TEXT NOT NULL,
    device_public_key   TEXT NOT NULL DEFAULT ''
);
-- Defensive ADD COLUMN for databases created before semp-go v0.4.0 added
-- DevicePublicKey to the DeviceCertificate struct. SQLite ignores
-- duplicates of an already-present column at this guard's level so the
-- statement is safe to re-run on every startup; we tolerate the error
-- raised on existing schemas via the InitDB wrapper.

CREATE TABLE IF NOT EXISTS block_entries (
    id           TEXT NOT NULL PRIMARY KEY,
    user_id      TEXT NOT NULL,
    entity_type  TEXT NOT NULL,
    entity_value TEXT NOT NULL,
    acknowledgment TEXT NOT NULL DEFAULT 'rejected',
    reason       TEXT,
    scope        TEXT NOT NULL DEFAULT 'all',
    created_at   TEXT NOT NULL,
    expires_at   TEXT
);
CREATE INDEX IF NOT EXISTS idx_block_user ON block_entries(user_id);

CREATE TABLE IF NOT EXISTS inbox (
    id        INTEGER PRIMARY KEY AUTOINCREMENT,
    address   TEXT NOT NULL,
    payload   BLOB NOT NULL,
    stored_at TEXT NOT NULL DEFAULT (datetime('now'))
);
CREATE INDEX IF NOT EXISTS idx_inbox_address ON inbox(address);

CREATE TABLE IF NOT EXISTS delivered_ids (
    address      TEXT NOT NULL,
    envelope_id  TEXT NOT NULL,
    delivered_at TEXT NOT NULL,
    PRIMARY KEY (address, envelope_id)
);
`

// InitDB opens or creates a SQLite database at path, applies the schema,
// and returns the handle. The database uses WAL mode for concurrent reads.
func InitDB(path string) (*sql.DB, error) {
	db, err := sql.Open("sqlite", path)
	if err != nil {
		return nil, fmt.Errorf("store: open %s: %w", path, err)
	}
	for _, pragma := range []string{
		"PRAGMA journal_mode=WAL",
		"PRAGMA foreign_keys=ON",
		"PRAGMA busy_timeout=5000",
	} {
		if _, err := db.Exec(pragma); err != nil {
			db.Close()
			return nil, fmt.Errorf("store: %s: %w", pragma, err)
		}
	}
	if _, err := db.Exec(schemaSQL); err != nil {
		db.Close()
		return nil, fmt.Errorf("store: schema init: %w", err)
	}
	// Apply post-schema migrations. ALTER TABLE ADD COLUMN is the
	// minimal-disruption way to grow existing databases; SQLite raises a
	// "duplicate column name" error when the column already exists, which
	// we swallow.
	for _, alter := range []string{
		`ALTER TABLE device_certificates ADD COLUMN device_public_key TEXT NOT NULL DEFAULT ''`,
	} {
		if _, err := db.Exec(alter); err != nil && !isDuplicateColumnErr(err) {
			db.Close()
			return nil, fmt.Errorf("store: migration %q: %w", alter, err)
		}
	}
	return db, nil
}

// isDuplicateColumnErr returns true when err is the SQLite error raised
// for an ALTER TABLE ADD COLUMN whose column already exists. The error
// text is matched by substring because modernc.org/sqlite does not
// surface the underlying SQLITE_ERROR code as a typed constant.
func isDuplicateColumnErr(err error) bool {
	if err == nil {
		return false
	}
	msg := err.Error()
	return contains(msg, "duplicate column name") || contains(msg, "already exists")
}

func contains(haystack, needle string) bool {
	if len(needle) == 0 || len(haystack) < len(needle) {
		return needle == ""
	}
	for i := 0; i+len(needle) <= len(haystack); i++ {
		if haystack[i:i+len(needle)] == needle {
			return true
		}
	}
	return false
}

File: impl/ts/src/store/schema.ts
/**
 * SQLite schema loader. Reads shared/schema/0001_init.sql so the TS
 * impl shares one source of truth with the Go impl.
 *
 * @module
 */

import { readFileSync } from "node:fs";

import type Database from "better-sqlite3";

const SCHEMA_REL_URL = new URL(
  "../../../../shared/schema/0001_init.sql",
  import.meta.url,
);

/** Apply pragmas, run the schema DDL, and record migration 1. */
export function applySchema(db: Database.Database): void {
  db.pragma("journal_mode = WAL");
  db.pragma("foreign_keys = ON");
  db.pragma("busy_timeout = 5000");

  const sql = readFileSync(SCHEMA_REL_URL, "utf8");
  db.exec(sql);

  // Tracking table for applied migrations. The Go impl creates this
  // alongside the first apply; mirror that here.
  db.exec(
    `CREATE TABLE IF NOT EXISTS schema_migrations (version INTEGER PRIMARY KEY, applied_at TEXT)`,
  );
  const has = db
    .prepare<[number], { version: number }>(
      `SELECT version FROM schema_migrations WHERE version = ?`,
    )
    .get(1);
  if (has === undefined) {
    db.prepare(
      `INSERT INTO schema_migrations (version, applied_at) VALUES (1, datetime('now'))`,
    ).run();
  }

  // Defensive ALTER for databases pre-dating the device_public_key
  // column. The Go impl swallows the duplicate-column error; mirror
  // that here.
  try {
    db.exec(
      `ALTER TABLE device_certificates ADD COLUMN device_public_key TEXT NOT NULL DEFAULT ''`,
    );
  } catch (err) {
    if (!isDuplicateColumnError(err)) {
      throw err;
    }
  }
}

function isDuplicateColumnError(err: unknown): boolean {
  const msg = err instanceof Error ? err.message : String(err);
  return msg.includes("duplicate column name") || msg.includes("already exists");
}
